source/common/network/tcp_keepalive_option_impl.cc

```diff
@@ -0,0 +1,76 @@
+#include "common/network/tcp_keepalive_option_impl.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+bool TcpKeepaliveOptionImpl::setOption(Network::Socket& socket,
+                                       Network::Socket::SocketState state) const {
+  if (state == Socket::SocketState::PreBind) {
+    return setTcpKeepalive(socket, keepalive_config_.keepalive_probes_,
+                           keepalive_config_.keepalive_time_,
+                           keepalive_config_.keepalive_interval_);
+  }
+  return true;
+}
+
+bool TcpKeepaliveOptionImpl::setTcpKeepalive(Socket& socket,
+                                             const absl::optional<uint32_t>& keepalive_probes,
+                                             const absl::optional<uint32_t>& keepalive_time,
+                                             const absl::optional<uint32_t>& keepalive_interval) {
+  if (!ENVOY_SOCKET_SO_KEEPALIVE.has_value()) {
+    ENVOY_LOG(warn, "TCP keepalive not supported on this platform");
+    return false;
+  }
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+
+  const int optOne = 1;
+  const socklen_t optOneLen = sizeof(optOne);
+  int error = os_syscalls.setsockopt(socket.fd(), SOL_SOCKET, ENVOY_SOCKET_SO_KEEPALIVE.value(),
+                                     &optOne, optOneLen);
+  if (error != 0) {
+    ENVOY_LOG(warn, "Setting SO_KEEPALIVE on socket failed: {}", strerror(error));
+    return false;
+  }
+  if (keepalive_probes.has_value()) {
```

Comment:
Please factor out a function for the pattern "if option doesn't exist then return false; set the option; check the error" and call it 4 times.